SingleLinkedList.java

```diff
@@ -0,0 +1,144 @@
+////////////////////////////////////////////////////////////////////////////////
+// Author: Ben Kuschner
+////////////////////////////////////////////////////////////////////////////////
+
+public class SingleLinkedList<T> {
+   
+   /////////////////////////////////////////////////////////
+   // Private link class
+   /////////////////////////////////////////////////////////
+   private class Link<T> {
+      // Member Variables
+      private T data;
+      private Link<T> next;
+
+      // Constructor(s)
+      public Link (T data) {
+         this.data = data;
+         next = null;
+      }
+   
+      // Member methods
+      public T getData() {
+         return this.data;
+      }
+      
+      public boolean hasNext() {
+    	  return next != null;
+      }
+   
+      public Link<T> getNext() {
+         return next;
+      }
+
+      public void setNext(Link<T> next) {
+         this.next = next;
+      }
+
+   } // end of class link
+
+   /////////////////////////////////////////////////////////
+   // Single Linked List
+   /////////////////////////////////////////////////////////
+
+   // Member variables
+   private Link<T> first;
+	
+   // Constructor(s)
+   public SingleLinkedList() {
+      first = null;
+   }
+
+   // Member methods
+	
+   public boolean isEmpty () {
+      return first == null;
+   }
+	
+   // Inserts a node at the front of the list.
+   public void insertFront(T data) {
+      Link<T> newLink = new Link<T>(data);
+
+      newLink.setNext(first);
+      first = newLink;
+   }
+
+   // Inserts a node at the back of the list
+   public void insertBack(T data) {
+      Link<T> last = first;
+      while (last.hasNext()) { //sets last to the last node
+    	  last = last.getNext();
+      }
+      last.setNext(new Link<T>(data));
+   }
+
+   // Returns the amount of objects in the list.
+   public int length() {
+      Link<T> location = first;
+      int length = 0;
+      while (location != null) {
+         length++;
+         location = location.getNext();
+      }
+      return length;
+   }
+	
+   // Removes the first node and returns the value stores in it.
+   public T popFront() {
+      T popped = first.getData();
+      first = first.getNext();
+      return popped;
+   }
+
+   // Removes the last node and returns the value stores in it.
+   public T popBack() {
```

Comment:
Once again it is probably worth it to store a reference to the last element in the list so that you do not need to linearly scan through the entire thing each time.

Ideally this method would be O(1), not O(n).